### PR TITLE
Fix max tokens constant

### LINKED ISSUE #352: Set `LLM_MAX_TOKENS` dynamically based on model selected
See https://github.com/cpacker/MemGPT/pull/349#issuecomment-1797857440

### PR DESCRIPTION
**Pull request by cpacker**
Closes #352, #353

**Please describe the purpose of this pull request.**
Patch #352

This PR touches:

- agent configs (test save/load)
- local LLM calls (test basic completion)
  - make sure that local LLMs get context length set to 8k by default, and that eg gpt-4-turbo gets set to 128k
- openai LLM calls (implicitly) (test openai functionality)
- summarizer calls (test summarize)

**How to test**

- With OpenAI: try `memgpt configure`, and it should autoselect the context window length
  - Then check that it's set correctly in config
- With local models: try `memgpt configure`, and it should ask for the context window length
  - Then check that it's set correctly in config
- With local models: try `memgpt run --context-window <N>`, and it should override the context window length
  - Check in config
  - But also test via local inference

**Have you tested this PR?**
Yes

**Related issues or PRs**
#352, #353

**Is your PR over 500 lines of code?**
n/a

**Additional context**
n/a

### CODE REVIEW

**Review by cpacker — COMMENTED** [review]
```
step() failed
user_message = None
error = 'NoneType' object has no attribute 'context_length'
'NoneType' object has no attribute 'context_length'
An exception ocurred when running agent.step(): 
Traceback (most recent call last):
  File "/Users/loaner/dev/MemGPT-public/MemGPT/memgpt/main.py", line 613, in run_agent_loop
    new_messages, user_message, skip_next_user_input = await process_agent_step(user_message, no_verify)
  File "/Users/loaner/dev/MemGPT-public/MemGPT/memgpt/main.py", line 589, in process_agent_step
    new_messages, heartbeat_request, function_failed, token_warning = await memgpt_agent.step(
  File "/Users/loaner/dev/MemGPT-public/MemGPT/memgpt/agent.py", line 1157, in step
    raise e
  File "/Users/loaner/dev/MemGPT-public/MemGPT/memgpt/agent.py", line 1085, in step
    context_length=self.config.context_length,
AttributeError: 'NoneType' object has no attribute 'context_length'
```

File: memgpt/agent.py [hunk 1]
```diff
@@ -23,7 +23,8 @@
     MAX_PAUSE_HEARTBEATS,
     MESSAGE_CHATGPT_FUNCTION_MODEL,
     MESSAGE_CHATGPT_FUNCTION_SYSTEM_MESSAGE,
-    MESSAGE_SUMMARY_WARNING_TOKENS,
+    MESSAGE_SUMMARY_WARNING_FRAC,
+    LLM_MAX_TOKENS,
```
Comment (vivi) [inline memgpt/agent.py:27]: This isn't used?

File: memgpt/config.py [hunk 2]
```diff
@@ -51,6 +51,7 @@ class MemGPTConfig:
     # provider: str = "openai"  # openai, azure, local (TODO)
     model_endpoint: str = "openai"
     model: str = "gpt-4"  # gpt-4, gpt-3.5-turbo, local
+    context_window: int = LLM_MAX_TOKENS[model]
```
Comment (sarahwooders) [inline memgpt/config.py:54]: you might want to check to make sure `model` is a key

File: memgpt/config.py [hunk 3]
```diff
@@ -106,6 +107,7 @@ def load(cls) -> "MemGPTConfig":
 
             # read config values
             model = config.get("defaults", "model")
+            context_window = config.get("defaults", "context_window")
```
Comment (vivi) [inline memgpt/config.py:110]: backwards compat for older agent configs:
`config.get("defaults", "context_window") if config.has_option("defaults", "context_window") else <some default value>`

File: memgpt/cli/cli_config.py [hunk 4]
```diff
@@ -102,6 +103,9 @@ def configure():
     else:
         default_model = "local"  # TODO: figure out if this is ok? this is for local endpoint
 
+    # get the max tokens (context window) for the model
+    default_model_context_window = LLM_MAX_TOKENS[default_model] if default_model in LLM_MAX_TOKENS else LLM_MAX_TOKENS["DEFAULT"]
```
Comment (sarahwooders) [inline memgpt/cli/cli_config.py:107]: don't you want them to specify the token count for the model if its not in the dict?

**Comment by sarahwooders** [comment]
You should maybe also modify this https://github.com/cpacker/MemGPT/blob/main/memgpt/cli/cli.py#L121 -- do you want to pass in token as a flag to `memgpt run` or only have it setup in `memgpt configure`?

**Comment by cpacker** [comment]
> You should maybe also modify this https://github.com/cpacker/MemGPT/blob/main/memgpt/cli/cli.py#L121 -- do you want to pass in token as a flag to `memgpt run` or only have it setup in `memgpt configure`?

I think for local llms it makes sense to pass it via `memgpt run` too

**Comment by cpacker** [comment]
Also @sarahwooders any idea about the runtime error? Seems like `agent.config` is None at the time the first `agent.step` is called

**Comment by vivi** [comment]
> Also @sarahwooders any idea about the runtime error? Seems like `agent.config` is None at the time the first `agent.step` is called

Looks like this is the case when the legacy cli is run: https://github.com/cpacker/MemGPT/blob/main/memgpt/main.py#L347

Maybe we should pass in a mock agent config

File: memgpt/cli/cli_config.py (hunk 4 above)
Comment (cpacker) [reply memgpt/cli/cli_config.py]: Changed to select from defaults if not openai:

image

File: memgpt/agent.py [hunk 5]
```diff
@@ -582,7 +580,12 @@ def step(self, user_message, first_message=False, first_message_retry_limit=FIRS
                 printd(f"This is the first message. Running extra verifier on AI response.")
                 counter = 0
                 while True:
-                    response = get_ai_reply(model=self.model, message_sequence=input_message_sequence, functions=self.functions)
+                    response = get_ai_reply(
+                        model=self.model,
+                        message_sequence=input_message_sequence,
+                        functions=self.functions,
+                        context_window=self.config.context_window,
```
Comment (cpacker) [inline memgpt/agent.py:587]: @sarahwooders example of grabbing the config inside of the `Agent` instance, is this correct?

Comment (cpacker) [reply memgpt/agent.py:587]: `[agent/self].config.context_window`

File: memgpt/agent.py [hunk 6]
```diff
@@ -620,14 +628,18 @@ def step(self, user_message, first_message=False, first_message_retry_limit=FIRS
             # Check the memory pressure and potentially issue a memory pressure warning
             current_total_tokens = response["usage"]["total_tokens"]
             active_memory_warning = False
-            if current_total_tokens > MESSAGE_SUMMARY_WARNING_TOKENS:
-                printd(f"WARNING: last response total_tokens ({current_total_tokens}) > {MESSAGE_SUMMARY_WARNING_TOKENS}")
+            # if current_total_tokens > MESSAGE_SUMMARY_WARNING_TOKENS:
+            # if current_total_tokens > MESSAGE_SUMMARY_WARNING_FRAC * LLM_MAX_TOKENS[self.model]:
```
Comment (vivi) [inline memgpt/agent.py:632]: stray comments

File: memgpt/constants.py [hunk 7]
```diff
@@ -17,9 +17,28 @@
...
+    "gpt-3.5-turbo-1106": 16385,
+    "gpt-3.5-turbo": 4096,
+    "gpt-3.5-turbo-16k": 16385,
+    "gpt-3.5-turbo-0613": 4096,  # legacy
+    "gpt-3.5-turbo-16k-0613": 16385,  # legacy
+    "gpt-3.5-turbo-0301": 4096,  # legacy
+}
 # The amount of tokens before a sytem warning about upcoming truncation is sent to MemGPT
-MESSAGE_SUMMARY_WARNING_TOKENS = int(0.75 * LLM_MAX_TOKENS)
+# MESSAGE_SUMMARY_WARNING_TOKENS = {k: int(0.75 * v) for k, v in LLM_MAX_TOKENS.items()}
```
Comment (vivi) [inline memgpt/constants.py:40]: stray comment?

**Review by vivi — APPROVED** [review]

**Review by sarahwooders — APPROVED** [review]
You might want to update docs with the `--context-window`flag, but fine to punt that to later.

**Pull request merged**